Inline `switch` special function call in rir

StrongerXi opened the pull request:
This PR attempts to solve #906 
Currently `switch` special function call is dispatched to the built-in implementation in R.
Inlining it in rir enables more analysis and optimizations in pir.

StrongerXi committed:
Draft for compiling switch in rir.

StrongerXi commented on rir/src/ir/Compiler.cpp:
```diff
@@ -1072,6 +1072,150 @@ bool compileSpecialCall(CompilerContext& ctx, SEXP ast, SEXP fun, SEXP args_,
+        /******************* INSTRUCTIONS START HERE *********************/
+        compileExpr(ctx, args[0]);
+        // the inlined list of types is based on definition of `isVector` in `Rinlinedfuns.h`
+        for (auto type : {LGLSXP, INTSXP, REALSXP, CPLXSXP, STRSXP,
```
I tried adding support for `BC::isType(TypeChecks::Vector)` (also for `String` and `Factor`), and handle it accordingly in `interp.cpp`. However I found that the `isType` instruction currently only comes from pir, so I decided to refrain from that and use `BC::is` instead.

StrongerXi commented:
Notice the `TODO`s in comments point out several places where we need to insert an `errorcall` instruction, which is currently not supported. (the current behavior is to either not inline the call, or return `Nil`. Maybe I should open up another PR to work on that, and then finish up this one?

o- commented:
this looks reasonable to me on the first glance. I didn't check the instructions you emit in detail, let me know if you get stuck on fixing some of the broken tests, or need help in reproducing them locally. it can be tricky at times.

for the error cases: you could use the same trick as on https://github.com/reactorlabs/rir/pull/920/files#diff-e7dea8cf4dd72c5c6e070769bc4adc22R1172 and use callBuiltin to call the error builtin, that should do the trick without adding any bytecodes.

StrongerXi commented:
I should've mentioned. I did try using `BC::callBuiltin`, but the `errorcall` function is not a builtin apparently, i.e. it's not in `R_FunTab`, and in R they only invoke it as a helper function in the implementations of other builtin.

o- commented:
> I should've mentioned. I did try using `BC::callBuiltin`, but the `errorcall` function is not a builtin apparently, i.e. it's not in `R_FunTab`, and in R they only invoke it as a helper function in the implementations of other builtin.

ah, you are way ahead of me :) try the `stop` internal

StrongerXi committed:
Fix bugs and update documentation

StrongerXi committed:
Add integeration tests for switch

StrongerXi committed:
Add meta info for BC::eq to make CodeVerifier happy. Leverage istype to call isFactor from R and fix extra warning from switch

StrongerXi committed:
Temporarily remove 'isFactor' check

StrongerXi committed:
Completely remove things added for isFactor check. Fix stack alighment issue. Compiling sub-expr that contains 'next' could break stack alignment.

StrongerXi committed:
Remove debug printing

StrongerXi commented:
Rebasing onto master to resolve travis ci build failure.

StrongerXi committed:
Attempt to fix travis build failure by explictly linking libintl

StrongerXi committed:
Rebase onto master. Travis build failure is fixed by a newly merged PR

o- commented:
@StrongerXi hey, what is the status? I see there is still a todo in there. can I help with something? would be cool if we can merge it soon!

StrongerXi commented:
Hi @o- . 

There are 2 todos:
1. A corner case that involves coercing `RAWSXP` to `INTSXP`. Handling this would probably require adding an `asInteger` instruction. The corner case hasn't shown up in any test, but I can try adding that instruction if you deem it necessary.
2. I couldn't simulate the `isFactor` check. I tried (a) using the builtin `inherit` function, (b) adding `TypeCHecks::Factor` and handle it in `BC::isType`. Both fail with different degree of errors, as documented in the code.

Currently only Travis fails with a time out error, but I'm not sure how to resolve that.

o- commented:
the test that we have are really not that extensive. so if you know about issue, we should fix them and if possible add tests that check them. it's unfortunately easy to pass the tests with broken things...

if you give me examples of switches that still fail I can help to debug, I have some free time again...

o- committed:
Merge branch 'master' into compile-switch

o- committed:
cleanup typechecks and add missing factor check

o- committed:
add constantfoldings

o- committed:
typo

o- committed:
fix updatePromise

Update promise is only responsible for updating escaped promises.
We must not update non-escaped promises.

Additionally keep track of updated promises in scope analysis.

o- committed:
fix controlflow issue after "stop"

o- committed:
it is possible to have a push context without pop

this happens due to unreachable code, e.g. stop() or similar

o- committed:
Merge branch 'forceDominanceUpdatePromiseFix2' into compile-switch
